### DISCUSSION TITLE
Rehydration & Actions

### DISCUSSION THREAD

**Discussion by jhrr** [statelyai/xstate#4671] [Q&A]
Hello, hello.

I'm just seeking some clarity here as I can't quite seem to locate it in the docs or in any of the discussions here.

I'm using v5 (really great work) and rehydrating a machine from localstorage using the snapshot parameter with a snapshot derived from `actor.getPersistedSnapshot()`.  ie.

`createActor(someMachine, { snapshot: someSnapshot });`

This works fine. I am able to rehydrate my machine back to the correct state with everything intact.

However, what I can't seem to tell is whether or not the actions should or shouldn't fire on the event of rehydration.

In my case I have an entry action on a state that, when it's rehydrated into that state, I need to fire again. As if the rehydration itself were entering the state anew.

However it doesn't seem to happen and I can't tell from the docs whether it should happen or even if it's possible.

I can see conceptually why this wouldn't or shouldn't happen. Something like: the state has been entered in the past and we are basically caching that fact when we dehydrate the machine, so don't refire things when it's brought back from storage.

However, in this particular case it would be very convenient if I could tell the machine to refire those actions.

I initially thought I might be able to mark the entire state with `reenter: true` but it looks like I can only mark actual transitions within an `on` block as reentrant in this way.

Am I missing something or is this just not something I can do without having to send an explicit event to the newly rehydrated machine to get actions firing again?

Many thanks for all the effort spent on the library, it's been life changing.

**Comment by Andarist** [comment]
> However, what I can't seem to tell is whether or not the actions should or shouldn't fire on the event of rehydration.

The current model of our rehydration is that it's like a resume after a pause. So not actions should be fired. In fact, actions can depend on the current `event` of the transition that led to their execution but when you rehydrate there is no transition taking place and thus no event to use there. 

> In my case I have an entry action on a state that, when it's rehydrated into that state, I need to fire again. As if the rehydration itself were entering the state anew.

It would be great to hear more about the actual concrete use case that you have there. 

---

You could try to imitate the desired behavior by using `invoke: { src: fromCallback(() => { /* my entry side-effects here */ }) }`.

**Reply by jhrr** [reply]
> You could try to imitate the desired behavior by using invoke: { src: fromCallback(() => { /* my entry side-effects here */ }) }.

This solves my use case. Great stuff thanks for that pointer.

> It would be great to hear more about the actual concrete use case that you have there.

I'm writing an OAuth flow machine with an OIDC/PKCE component for an SPA and serializing that machine to storage for persistence.

This is a bit simplified for illustration but basically, if you've authorized yourself with the system at some time in the past, the machine, when it's rehydrated on page load, might be in the `authorized` state. So far so good, but to mitigate certain vulnerabilities around SPAs and JWT tokens I don't allow the access token to be serialized along with the machine, I just keep it in a reactive ref so it lasts as long as the tab lifetime and then I keep the refresh token in an HTTP only cookie. However, some less sensitive data like the OIDC id token and the authenticated user data I do keep in the machine context which gets dehydrated and persisted to `localstorage`.

So, there are basically two pathways into `authorized`. Either from fully logging in, in which case there will be an access token available and that's that, or from rehydrating the machine which might be in the `authorized` state, in which case there won't be an access token available but then I can check for data in the context or whatever and just assume I can get a new one using the refresh token in the cookie (unless that has expired in which case we can transition to `unauthorized` or wherever else).

In other words, if I can check the access token in the entry action to `authorized` I can stop there if that's available. But if the access token is `null` and the machine is in `authorized` I can transition to the `refresh` state, get a new access token and finally return to `authorized` and everything can resume. But obviously this approach wont work if nothing fires on rehydration! We'll be in the `authorized` state but we can't get anything from the API as we don't have any credentials.

That's basically it, but your solution allows me to make that token check even on the event of rehydration, and this all works correctly now.

Cheers!
